[SR-15940] Assert crash in TypeCheckStmt.cpp

|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15940      |
   |Radar             | None         |
   |Original Reporter | ellis (JIRA User)      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       |     |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   md5: c0318b50fadbb361c90032f67b6a66fb

  </details>

**Issue Description:**
Swift crashes when compiling the following source.

```swift
@propertyWrapper
public class Bar<Value> {
  public init(wrappedValue: @autoclosure () -> Value) {}
  public var wrappedValue: Value {}
}

class Foo {
  @Bar var a: Bool?
}
```

Here is the backtrace

```
error: compile command failed due to signal 6 (use -v to see invocation)
test.swift:4:35: error: computed property must have accessors specified
public var wrappedValue: Value {}
^
swift-frontend: swift/lib/Sema/TypeCheckStmt.cpp:86: virtual std::pair<bool, Expr *> (anonymous namespace)::ContextualizeClosures::walkToExprPre(swift::Expr *): Assertion `CE->getDiscriminator() == AutoClosureExpr::InvalidDiscriminator' failed.
Please submit a bug report (<https://swift.org/contributing/#reporting-bugs)> and include the project and the crash backtrace.
Stack dump:
0. Program arguments: swift-frontend -frontend -c -primary-file test.swift -target arm64-apple-ios11.0 -Xllvm -aarch64-use-tbi -enable-objc-interop -color-diagnostics -new-driver-path /swift-driver -resource-dir swift -module-name EmojiPickerLayout -target-sdk-version 15.2 -o /tmp/TemporaryDirectory.qXz2A9/EmojiPickerLayout-1.o
1. Swift version 5.6-dev (LLVM a5fb80477a008fd, Swift c2abee2cb7c2471)
2. Compiling with the current language version
3. While evaluating request TypeCheckSourceFileRequest(source_file "test.swift")
4. While type-checking 'Foo' (at test.swift:7:1)
5. While type-checking declaration 0x8841d80 (in module 'EmojiPickerLayout')
#0 0x0000000007d136f3 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) llvm-project/llvm/lib/Support/Unix/Signals.inc:565:13
#1 0x0000000007d11a50 llvm::sys::RunSignalHandlers() llvm-project/llvm/lib/Support/Signals.cpp:98:18
#2 0x0000000007d13a7f SignalHandler(int) llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
#3 0x00007f72a1d3f130 __restore_rt (/usr/local/fbcode/platform009/lib/libpthread.so.0+0x15130)
#4 0x00007f72a17d2530 raise /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
#5 0x00007f72a17b951d abort /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/stdlib/abort.c:81:7
#6 0x00007f72a17c94c3 __assert_fail_base /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/assert/assert.c:89:7
#7 0x00007f72a17c9533 (/usr/local/fbcode/platform009/lib/libc.so.6+0x35533)
#8 0x00000000040db1cf swift::AbstractClosureExpr::setDiscriminator(unsigned int) swift/include/swift/AST/Expr.h:3610:5
#9 0x00000000040db1cf (anonymous namespace)::ContextualizeClosures::walkToExprPre(swift::Expr*) swift/lib/Sema/TypeCheckStmt.cpp:87:13
#10 0x00000000043404af (anonymous namespace)::Traversal::doIt(swift::Expr*) swift/lib/AST/ASTWalker.cpp:1197:14
#11 0x00000000043404af (anonymous namespace)::Traversal::doIt(swift::ArgumentList*) swift/lib/AST/ASTWalker.cpp:1360:17
#12 0x00000000043432cd (anonymous namespace)::Traversal::visitApplyExpr(swift::ApplyExpr*) swift/lib/AST/ASTWalker.cpp:856:15
#13 0x000000000433f891 (anonymous namespace)::Traversal::SetParentRAII::~SetParentRAII() swift/lib/AST/ASTWalker.cpp:93:21
#14 0x000000000433f891 (anonymous namespace)::Traversal::visit(swift::Expr*) swift/lib/AST/ASTWalker.cpp:100:3
#15 0x000000000433f891 (anonymous namespace)::Traversal::doIt(swift::Expr*) swift/lib/AST/ASTWalker.cpp:1201:9
#16 0x000000000433f891 swift::Expr::walk(swift::ASTWalker&) swift/lib/AST/ASTWalker.cpp:1891:28
#17 0x00000000040d64ab swift::TypeChecker::contextualizeInitializer(swift::Initializer*, swift::Expr*) swift/lib/Sema/TypeCheckStmt.cpp:202:1
#18 0x0000000004024715 (anonymous namespace)::DeclChecker::visitPatternBindingDecl(swift::PatternBindingDecl*) swift/lib/Sema/TypeCheckDeclPrimary.cpp:2088:13
#19 0x0000000004024715 swift::ASTVisitor<(anonymous namespace)::DeclChecker, void, void, void, void, void, void>::visit(swift::Decl*) swift/include/swift/AST/DeclNodes.def:185:1
#20 0x0000000004024715 (anonymous namespace)::DeclChecker::visit(swift::Decl*) swift/lib/Sema/TypeCheckDeclPrimary.cpp:1697:31
#21 0x000000000402a1ac (anonymous namespace)::DeclChecker::visitClassDecl(swift::ClassDecl*) swift/lib/Sema/TypeCheckDeclPrimary.cpp:2478:23
#22 0x0000000004023aeb swift::ASTVisitor<(anonymous namespace)::DeclChecker, void, void, void, void, void, void>::visit(swift::Decl*) swift/include/swift/AST/DeclNodes.def:0:9
#23 0x0000000004023aeb (anonymous namespace)::DeclChecker::visit(swift::Decl*) swift/lib/Sema/TypeCheckDeclPrimary.cpp:1697:31
#24 0x000000000402383e swift::TypeChecker::typeCheckDecl(swift::Decl*, bool) swift/lib/Sema/TypeCheckDeclPrimary.cpp:3186:1
#25 0x000000000410e267 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const swift/lib/Sema/TypeChecker.cpp:336:17
#26 0x0000000004110258 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) swift/include/swift/AST/Evaluator.h:362:14
#27 0x000000000410ff72 llvm::Expected<std::tuple<> >::operator bool() llvm-project/llvm/include/llvm/Support/Error.h:564:17
#28 0x000000000410ff72 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) swift/include/swift/AST/Evaluator.h:393:10
#29 0x000000000410e0c4 llvm::Expected<std::tuple<> >::takeError() llvm-project/llvm/include/llvm/Support/Error.h:592:15
#30 0x000000000410e0c4 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) swift/include/swift/AST/Evaluator.h:106:25
#31 0x000000000410e06e swift::performTypeChecking(swift::SourceFile&) swift/lib/Sema/TypeChecker.cpp:286:3
#32 0x000000000321ab3a swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<void (swift::SourceFile&)>) swift/lib/Frontend/Frontend.cpp:1138:19
#33 0x000000000321aa08 _*gnu_cxx::normal_iterator<unsigned int const*, std::vector<unsigned int, std::allocator<unsigned int> > >::*_normal_iterator(unsigned int const* const&) /mnt/gvfs/third-party2/libgcc/9ec50d4571ff7613c2a4dc750831a6d1cb86995f/9.x/platform009/9202ce7/include/c++/9.x/bits/stl_iterator.h:808:20
#34 0x000000000321aa08 std::vector<unsigned int, std::allocator<unsigned int> >::begin() const /mnt/gvfs/third-party2/libgcc/9ec50d4571ff7613c2a4dc750831a6d1cb86995f/9.x/platform009/9202ce7/include/c++/9.x/bits/stl_vector.h:818:16
#35 0x000000000321aa08 std::vector<unsigned int, std::allocator<unsigned int> >::empty() const /mnt/gvfs/third-party2/libgcc/9ec50d4571ff7613c2a4dc750831a6d1cb86995f/9.x/platform009/9202ce7/include/c++/9.x/bits/stl_vector.h:1005:16
#36 0x000000000321aa08 llvm::SetVector<unsigned int, std::vector<unsigned int, std::allocator<unsigned int> >, llvm::DenseSet<unsigned int, llvm::DenseMapInfo<unsigned int> > >::empty() const llvm-project/llvm/include/llvm/ADT/SetVector.h:73:20
#37 0x000000000321aa08 swift::CompilerInstance::isWholeModuleCompilation() const swift/include/swift/Frontend/Frontend.h:454:67
#38 0x000000000321aa08 swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<void (swift::SourceFile&)>) swift/lib/Frontend/Frontend.cpp:1129:7
#39 0x000000000321aa08 swift::CompilerInstance::finishTypeChecking() swift/lib/Frontend/Frontend.cpp:1145:3
#40 0x000000000321aa08 swift::CompilerInstance::performSema() swift/lib/Frontend/Frontend.cpp:1070:3
#41 0x000000000308123b withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>) swift/lib/FrontendTool/FrontendTool.cpp:1108:7
#42 0x000000000307838d performAction(swift::CompilerInstance&, int&, swift::FrontendObserver*) swift/lib/FrontendTool/FrontendTool.cpp:0:12
#43 0x000000000307838d performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) swift/lib/FrontendTool/FrontendTool.cpp:1306:19
#44 0x00000000030778f6 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) swift/lib/FrontendTool/FrontendTool.cpp:2080:46
#45 0x0000000003045f83 swift::mainEntry(int, char const**) swift/lib/DriverTool/driver.cpp:405:12
#46 0x00007f72a17b9dc5 __libc_start_main /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/csu/../csu/libc-start.c:308:16
#47 0x000000000304590a _start /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/csu/../sysdeps/x86_64/start.S:122:0
```

--- a/lib/Sema/TypeCheckStorage.cpp
+++ b/lib/Sema/TypeCheckStorage.cpp
@@ -2690,7 +2690,7 @@ static VarDecl *synthesizePropertyWrapperProjectionVar(
 
 static void typeCheckSynthesizedWrapperInitializer(VarDecl *wrappedVar,
                                                    Expr *&initializer,
-                                                   bool contextualize = false) {
+                                                   bool contextualize) {
   auto *dc = wrappedVar->getInnermostDeclContext();
   auto &ctx = wrappedVar->getASTContext();
   auto *initContext = new (ctx) PropertyWrapperInitializer(
@@ -2985,7 +2985,8 @@ PropertyWrapperInitializerInfoRequest::evaluate(Evaluator &evaluator,
         && !wrapperInfo.defaultInit) {
       auto ty = parentPBD->getPattern(patternNumber)->getType();
       if (auto defaultInit = TypeChecker::buildDefaultInitializer(ty)) {
-        typeCheckSynthesizedWrapperInitializer(var, defaultInit);
+        typeCheckSynthesizedWrapperInitializer(var, defaultInit,
+                                               /*contextualize=*/false);
         parentPBD->setInit(0, defaultInit);
         parentPBD->setInitializerChecked(0);
       }
@@ -3003,6 +3004,8 @@ PropertyWrapperInitializerInfoRequest::evaluate(Evaluator &evaluator,
         // FIXME: Record this expression somewhere so that DI can perform the
         // initialization itself.
         Expr *defaultInit = nullptr;
+        // Only contextualize local wrapped property, the rest of wrapped
+        // property will be contextualized in visitPatternBindingDecl.
         typeCheckSynthesizedWrapperInitializer(var, defaultInit, dc->isLocalContext());
         pbd->setInit(0, defaultInit);
         pbd->setInitializerChecked(0);
@@ -3073,7 +3076,8 @@ PropertyWrapperInitializerInfoRequest::evaluate(Evaluator &evaluator,
              !var->getName().hasDollarPrefix()) {
     wrappedValueInit = PropertyWrapperValuePlaceholderExpr::create(
         ctx, var->getSourceRange(), var->getType(), /*wrappedValue=*/nullptr);
-    typeCheckSynthesizedWrapperInitializer(var, wrappedValueInit, true);
+    typeCheckSynthesizedWrapperInitializer(var, wrappedValueInit,
+                                           /*contextualize=*/true);
   }
 
   return PropertyWrapperInitializerInfo(wrappedValueInit, projectedValueInit);

--- a/lib/Sema/TypeCheckStorage.cpp
+++ b/lib/Sema/TypeCheckStorage.cpp
@@ -2689,7 +2689,8 @@ static VarDecl *synthesizePropertyWrapperProjectionVar(
 }
 
 static void typeCheckSynthesizedWrapperInitializer(VarDecl *wrappedVar,
-                                                   Expr *&initializer) {
+                                                   Expr *&initializer,
+                                                   bool contextualize = false) {
   auto *dc = wrappedVar->getInnermostDeclContext();
   auto &ctx = wrappedVar->getASTContext();
   auto *initContext = new (ctx) PropertyWrapperInitializer(
@@ -2705,6 +2706,11 @@ static void typeCheckSynthesizedWrapperInitializer(VarDecl *wrappedVar,
 
   initializer = result->getAsExpr();
 
+  // Contextualize the initializer which is a local variable with defaultInit or
+  // gets an independent initializer. The rest of initializer contextualizing
+  // will be done in visitPatternBindingDecl.
+  if (!contextualize)
+    return;
   TypeChecker::contextualizeInitializer(initContext, initializer);
   checkPropertyWrapperActorIsolation(wrappedVar, initializer);
   TypeChecker::checkInitializerEffects(initContext, initializer);
@@ -2997,7 +3003,7 @@ PropertyWrapperInitializerInfoRequest::evaluate(Evaluator &evaluator,
         // FIXME: Record this expression somewhere so that DI can perform the
         // initialization itself.
         Expr *defaultInit = nullptr;
-        typeCheckSynthesizedWrapperInitializer(var, defaultInit);
+        typeCheckSynthesizedWrapperInitializer(var, defaultInit, dc->isLocalContext());
         pbd->setInit(0, defaultInit);
         pbd->setInitializerChecked(0);
 
@@ -3067,7 +3073,7 @@ PropertyWrapperInitializerInfoRequest::evaluate(Evaluator &evaluator,
              !var->getName().hasDollarPrefix()) {
     wrappedValueInit = PropertyWrapperValuePlaceholderExpr::create(
         ctx, var->getSourceRange(), var->getType(), /*wrappedValue=*/nullptr);
-    typeCheckSynthesizedWrapperInitializer(var, wrappedValueInit);
+    typeCheckSynthesizedWrapperInitializer(var, wrappedValueInit, true);
   }
 
   return PropertyWrapperInitializerInfo(wrappedValueInit, projectedValueInit);

--- a/test/Concurrency/global_actor_inference.swift
+++ b/test/Concurrency/global_actor_inference.swift
@@ -415,12 +415,13 @@ actor WrapperActorBad2<Wrapped: Sendable> {
 struct WrapperWithMainActorDefaultInit {
   var wrappedValue: Int { fatalError() }
 
-  @MainActor init() {} // expected-note {{calls to initializer 'init()' from outside of its actor context are implicitly asynchronous}}
+  @MainActor init() {} // expected-note 2 {{calls to initializer 'init()' from outside of its actor context are implicitly asynchronous}}
 }
 
 actor ActorWithWrapper {
   @WrapperOnActor var synced: Int = 0
   // expected-note@-1 3{{property declared here}}
+  @WrapperWithMainActorDefaultInit var property: Int // expected-error {{call to main actor-isolated initializer 'init()' in a synchronous actor-isolated context}}
   func f() {
     _ = synced // expected-error{{main actor-isolated property 'synced' can not be referenced on a different actor instance}}
     _ = $synced // expected-error{{global actor 'SomeGlobalActor'-isolated property '$synced' can not be referenced on a different actor instance}}

--- a/test/SILGen/property_wrappers.swift
+++ b/test/SILGen/property_wrappers.swift
@@ -957,6 +957,34 @@ struct TestAutoclosureComposition {
   @Once @ObservedObject var model = Model()
 }
 
+@propertyWrapper
+struct SR_15940Foo {
+  var wrappedValue: Int { 0 }
+}
+
+struct SR_15940_C {
+  func a() {
+    @SR_15940Foo var b: Int
+  }
+}
+
+// CHECK-LABEL: sil hidden [ossa] @$s17property_wrappers10SR_15940_CV1ayyF : $@convention(method) (SR_15940_C) -> () {
+// CHECK: bb0(%0 : $SR_15940_C):
+// CHECK-NEXT:  debug_value %0 : $SR_15940_C, let, name "self", argno 1, implicit
+// CHECK-NEXT:  [[BOX:%.*]] = alloc_box ${ var SR_15940Foo }, var, name "_b"
+// CHECK-NEXT:  [[LIFETIME:%.*]] = begin_borrow [lexical] [[BOX]] : ${ var SR_15940Foo }
+// CHECK-NEXT:  [[BOXADDR:%.*]] = project_box [[LIFETIME]] : ${ var SR_15940Foo }, 0
+// CHECK-NEXT:  [[METATYPE:%.*]] = metatype $@thin SR_15940Foo.Type
+// CHECK-NEXT:  // function_ref SR_15940Foo.init()
+// CHECK-NEXT:  [[DEFAULTVALUE_FN:%.*]] = function_ref @$s17property_wrappers11SR_15940FooVACycfC : $@convention(method) (@thin SR_15940Foo.Type) -> SR_15940Foo
+// CHECK-NEXT:  [[DEFAULTRESULT:%.*]] = apply [[DEFAULTVALUE_FN]]([[METATYPE]]) : $@convention(method) (@thin SR_15940Foo.Type) -> SR_15940Foo
+// CHECK-NEXT:  store [[DEFAULTRESULT]] to [trivial] [[BOXADDR]] : $*SR_15940Foo
+// CHECK-NEXT:  end_borrow [[LIFETIME]] : ${ var SR_15940Foo }
+// CHECK-NEXT:  destroy_value [[BOX]] : ${ var SR_15940Foo }
+// CHECK-NEXT:  [[TUPLE:%.*]] = tuple ()
+// CHECK-NEXT:  return [[TUPLE]] : $()
+// CHECK-NEXT:  } // end sil function '$s17property_wrappers10SR_15940_CV1ayyF
+
 // CHECK-LABEL: sil_vtable ClassUsingWrapper {
 // CHECK-NEXT:  #ClassUsingWrapper.x!getter: (ClassUsingWrapper) -> () -> Int : @$s17property_wrappers17ClassUsingWrapperC1xSivg   // ClassUsingWrapper.x.getter
 // CHECK-NEXT:  #ClassUsingWrapper.x!setter: (ClassUsingWrapper) -> (Int) -> () : @$s17property_wrappers17ClassUsingWrapperC1xSivs // ClassUsingWrapper.x.setter

--- a/test/Sema/property_wrappers.swift
+++ b/test/Sema/property_wrappers.swift
@@ -47,3 +47,37 @@ struct TestInitSubscript {
   @Wrapper(wrappedValue: Color.allCases[0])
   var color: Color
 }
+
+@propertyWrapper
+public class SR_15940Bar<Value> {
+  private var _value: Value
+
+  public var wrappedValue: Value {
+    get { _value }
+    set {
+      _value = newValue
+    }
+  }
+
+  public init(wrappedValue value: @autoclosure @escaping () -> Value) {
+    self._value = value()
+  }
+}
+
+// CHECK-LABEL: struct_decl{{.*}}SR_15940_A
+struct SR_15940_A {
+  // CHECK:      argument_list implicit labels=wrappedValue:
+  // CHECK-NEXT:   argument label=wrappedValue
+  // CHECK-NEXT:     autoclosure_expr implicit type='() -> Bool?' discriminator=0 captures=(<opaque_value> ) escaping
+  // CHECK:            autoclosure_expr implicit type='() -> Bool?' discriminator=1 escaping
+  @SR_15940Bar var a: Bool?
+}
+
+// CHECK-LABEL: struct_decl{{.*}}SR_15940_B
+struct SR_15940_B {
+  // CHECK:      argument_list implicit labels=wrappedValue:
+  // CHECK-NEXT:   argument label=wrappedValue
+  // CHECK-NEXT:     autoclosure_expr implicit type='() -> Bool' location={{.*}}.swift:[[@LINE+2]]:30 range=[{{.+}}] discriminator=0 captures=(<opaque_value> ) escaping
+  // CHECK:            autoclosure_expr implicit type='() -> Bool' location={{.*}}.swift:[[@LINE+1]]:30 range=[{{.+}}] discriminator=1 escaping
+  @SR_15940Bar var b: Bool = false
+}

